src/libraries/System.Text.Json/src/System/Text/Json/Serialization/PooledByteBufferWriter.cs

```diff
@@ -138,7 +139,12 @@ private void CheckAndResizeBuffer(int sizeHint)
             {
                 int growBy = Math.Max(sizeHint, _rentedBuffer.Length);
 
-                int newSize = checked(_rentedBuffer.Length + growBy);
+                int newSize = _rentedBuffer.Length + growBy;
+
+                if ((uint)newSize > MaxByteArrayLength)
+                {
+                    throw new JsonException();
+                }
```

Comment:
This is throwing if _rentedBuffer.Length + growBy is too large, but that doesn't necessarily mean that _rentedBuffer.Length + sizeHint is too large.

It also seems like this should be an OutOfMemoryException, rather than a JsonException.  That's what the runtime would throw if too large a value were passed when new'ing the array, and this case really isn't different.

Consider using logic closer to that of `List<T>.EnsureCapacity`.
https://github.com/dotnet/runtime/blob/4d3ec15a0e197836f46d4297a5c6ae9722ec26f7/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs#L405-L409

Reply:
@stephentoub List doesn't handle this specific scenario. It only validates the "auto-grow" length, but not "min" parameter.

If you run

    var a = new List<int>();
    a.Capacity = int.MaxValue;

It tries to allocate the array and throws an OutOfMemoryException.

But we are trying to throw a more user-friendly Exception, so I can validate the size and throw an OutOfMemoryException with a custom message. Something like:

    int growBy = Math.Max(sizeHint, _rentedBuffer.Length);

    int newSize = _rentedBuffer.Length + growBy;

    if ((uint)newSize > MaxByteArrayLength && growBy > sizeHint)
    {
        newSize = _rentedBuffer.Length + sizeHint;
    }

    if ((uint)newSize > MaxByteArrayLength)
    {
        //newSize = MaxByteArrayLength;
        ThrowHelper.ThrowJsonException_ResultExceededMaximumBufferSize();//OutOfMemory
    }

I have a question, GetSpan and GetMemory can return a buffer smaller than sizeHint? Because CheckAndResizeBuffer has a Assert:

    Debug.Assert(_rentedBuffer.Length - _index >= sizeHint);

So I can't set newSize to MaxByteArrayLength. But some other classes validates the buffer returned by GetMemory: https://github.com/dotnet/runtime/blob/81bf79fd9aa75305e55abe2f7e9ef3f60624a3a1/src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.cs#L1082

Reply:
> But we are trying to throw a more user-friendly Exception

Why isn't an OutOfMemoryException sufficient?

Reply:
You are right, OutOfMemory should be obvious enough

Reply:
> I have a question, GetSpan and GetMemory can return a buffer smaller than sizeHint? Because CheckAndResizeBuffer has a Assert:

No, GetSpan/GetMemory shouldn't return anything smaller than sizeHint.
https://docs.microsoft.com/en-us/dotnet/api/system.buffers.ibufferwriter-1.getspan?view=netcore-3.1#System_Buffers_IBufferWriter_1_GetSpan_System_Int32_

The `Utf8JsonWriter` is doing extra checks for it, since someone could implement an `IBufferWriter` incorrectly that doesn't honor the contract and pass that to the `public` API. In this case, `PooledBufferWriter` is internal and we ensure it honors the contract or throws.